lib/kubeclient/oidc_auth_provider.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+module Kubeclient
+  # Uses OIDC id-tokens and refreshes them if they are stale.
+  class OIDCAuthProvider
+    class OpenIDConnectDependencyError < LoadError # rubocop:disable Lint/InheritException
+    end
+
+    class << self
+      def token(auth_provider)
+        begin
+          require 'openid_connect'
+        rescue LoadError => e
+          raise OpenIDConnectDependencyError,
+                'Error requiring openid_connect gem. Kubeclient itself does not include the ' \
+                'openid_connect gem. To support auth-provider oidc, you must include it in your ' \
+                "calling application. Failed with: #{e.message}"
+        end
+
+        issuer_url = auth_provider['idp-issuer-url']
+        discovery = OpenIDConnect::Discovery::Provider::Config.discover! issuer_url
+
+        id_token = OpenIDConnect::ResponseObject::IdToken.decode auth_provider['id-token'],
```

Comment:
What if the config doesn't contain `id-token` at all?  (Is that possible?)
As we can always obtain a new one, let's allow this?

Reply:
I'm not sure how you'd get into that state but `kubectl` supports it so I've implemented it too.